metrics-tracing-context/src/tracing_integration.rs

```diff
@@ -1,28 +1,37 @@
 //! The code that integrates with the `tracing` crate.
 
+use indexmap::IndexMap;
 use lockfree_object_pool::{LinearObjectPool, LinearOwnedReusable};
 use metrics::{Key, Label};
 use once_cell::sync::OnceCell;
 use std::sync::Arc;
-use std::{any::TypeId, marker::PhantomData};
+use std::{any::TypeId, cmp, marker::PhantomData};
 use tracing_core::span::{Attributes, Id, Record};
 use tracing_core::{field::Visit, Dispatch, Field, Subscriber};
 use tracing_subscriber::{layer::Context, registry::LookupSpan, Layer};
 
-fn get_pool() -> &'static Arc<LinearObjectPool<Vec<Label>>> {
-    static POOL: OnceCell<Arc<LinearObjectPool<Vec<Label>>>> = OnceCell::new();
-    POOL.get_or_init(|| Arc::new(LinearObjectPool::new(Vec::new, Vec::clear)))
+pub(crate) type Map = IndexMap<&'static str, Label>;
+
+fn get_pool() -> &'static Arc<LinearObjectPool<Map>> {
+    static POOL: OnceCell<Arc<LinearObjectPool<Map>>> = OnceCell::new();
+    POOL.get_or_init(|| Arc::new(LinearObjectPool::new(Map::new, Map::clear)))
 }
+
 /// Span fields mapped as metrics labels.
 ///
 /// Hidden from documentation as there is no need for end users to ever touch this type, but it must
 /// be public in order to be pulled in by external benchmark code.
 #[doc(hidden)]
-pub struct Labels(pub LinearOwnedReusable<Vec<Label>>);
+pub struct Labels(pub LinearOwnedReusable<Map>);
 
 impl Labels {
     pub(crate) fn extend_from_labels(&mut self, other: &Labels) {
-        self.0.extend_from_slice(other.as_ref());
+        let new_len = cmp::max(self.as_ref().len(), other.as_ref().len());
+        let additional = new_len - self.as_ref().len();
+        self.0.reserve(additional);
+        for (k, v) in other.as_ref() {
+            self.0.insert(k, v.clone());
+        }
```

Comment:
So, the opposite of what's described here? https://github.com/metrics-rs/metrics/pull/408/files#diff-a16803449fe3229b496af705d5d434d1dee993cbd2606a6fb04f3a6a645a9955R71-R76 I thought it would be correct to overwrite them. Or do you mean this:

```rust
tracing::info_span!("span", field = 1).in_scope(|| {
	metrics::counter(
		"counter",
		"field" => 2,
	).increment(1); // would increment `counter{field=1}`, because value `2` is overwritten by the span
});
```

In that case, yeah, this should be done differently then.

Reply:
This is tricky, as we want to keep only the "most recent" value from spans, but discard it entirely if it clashes with one of the labels originally used...

Reply:
To address my comment on this code block specifically, though, what we currently do in `on_new_span` is:

```
let mut labels = Labels::from_record(&Record::new(attrs.values()));

if let Some(parent) = span.parent() {
    if let Some(parent_labels) = parent.extensions().get::<Labels>() {
        labels.extend_from_labels(parent_labels);
    }
}
```

We create our set of labels from the current span's attributes, and then we extend those labels from the parent span's calculated labels, if any. Based on how the code is written in `extend_from_labels`, any overlapping labels in the parent span's labels would overwrite our own labels from the current span.